widgetssdk/src/main/java/com/glia/widgets/call/CallActivity.java

```diff
@@ -145,9 +146,11 @@ private void startCallWithPermissions() {
             if (configuration.getMediaType() == Engagement.MediaType.VIDEO && missingPermission(Manifest.permission.CAMERA)) {
                 missingPermissions.add(Manifest.permission.CAMERA);
             }
-            if ((configuration.getMediaType() == Engagement.MediaType.VIDEO || configuration.getMediaType() == Engagement.MediaType.AUDIO)
-                    && missingPermission(Manifest.permission.RECORD_AUDIO)) {
-                missingPermissions.add(Manifest.permission.RECORD_AUDIO);
+            if (!Dependencies.getUseCaseFactory().createIsCallVisualizerUseCase().invoke()) {
+                if ((configuration.getMediaType() == Engagement.MediaType.VIDEO || configuration.getMediaType() == Engagement.MediaType.AUDIO)
+                        && missingPermission(Manifest.permission.RECORD_AUDIO)) {
+                    missingPermissions.add(Manifest.permission.RECORD_AUDIO);
+                }
```

Comment:
I prefer to move this to the CallController.
Also, I don't think it is a good idea to get the UseCase there.

Reply:
@Pelkar Agree with Andrii. 
et me also ask you, have you understood why the Call screen is opened without permission for the first time but asks permission the second time?

Reply:
> I prefer to move this to the CallController. Also, I don't think it is a good idea to get the UseCase there.

Good suggestion, will implement the changes

Reply:
> @Pelkar Agree with Andrii. Let me also ask you, have you understood why the Call screen is opened without permission for the first time but asks permission the second time?

It seems to me that the reason is pretty simple. Upgrade permissions are handled on the core side and core only requests camera with no audio(as it should). The resume that triggers the second dialog however wants to have all permissions a regular call engagement would need, thus audio dialog is asked separately when view resumes outside of upgrade

Reply:
I migrated the missing permission logic to controller. However, because there is no concrete way of communication from view structure to controller and back, the current solution is not as elegant as it could be...